modules/lib/lib-portal/src/main/resources/lib/xp/portal.ts

```diff
@@ -83,15 +83,70 @@ export interface SiteConfig<Config> {
     config: Config;
 }
 
-export interface AssetUrlParams {
+export interface Cookie {
+    value: string;
+    path?: string;
+    domain?: string;
+    maxAge?: number;
+    secure?: boolean;
+    httpOnly?: boolean;
+    sameSite?: 'lax' | 'strict' | 'none' | '';
+}
+
+type LiteralUnion<T extends U, U = string> = T | (U & Record<never, never>);
+
+export type RequestParams = Record<string, string | string[] | undefined>;
+
+export interface Request<Params extends RequestParams = RequestParams,
+    Headers extends Record<string, string | undefined> = Record<string, string | undefined>,
+    Cookies extends Record<string, string | Cookie | undefined> = Record<string, string | Cookie | undefined>,
+    > {
+    method: LiteralUnion<'GET' | 'PUT' | 'POST' | 'DELETE' | 'HEAD' | 'PATCH' | 'OPTIONS' | 'TRACE' | 'CONNECT'>;
+    scheme: 'http' | 'https';
+    host: string;
+    port: number;
+    path: string;
+    rawPath: string;
+    url: string;
+    remoteAddress: string;
+    mode: 'inline' | 'edit' | 'preview' | 'live' | 'admin';
+    webSocket: boolean;
+    repositoryId: string;
+    branch: LiteralUnion<'draft' | 'master'>;
+    contextPath: string;
+    body?: string;
+    params: Params;
+    headers: Headers;
+    cookies: Cookies;
+    contentType?: string;
+}
+
+export interface Response<ResponseBody = unknown> {
+    status: number;
+    body: ResponseBody;
+    contentType: string;
+    headers: Record<string, string | undefined>;
+    cookies: Record<string, string | Cookie | undefined>;
+    redirect: string;
```

Comment:
```suggestion
    redirect?: string;
```

redirect is a "write only" property, initially it is undefined